### ISSUE TITLE
inventory parsing performance issue with vars_plugins

### ISSUE THREAD

**Issue by renard** [ansible/ansible#19993]
##### ISSUE TYPE

-  Feature Idea
- ~~Bug Report~~
 
##### COMPONENT NAME

inventory parsing

##### ANSIBLE VERSION

```
ansible 2.2.0.0 (stable-2.2 4cd32ee1ac) last updated 2016/10/12 21:20:06 (GMT +200)
  lib/ansible/modules/core: (detached HEAD cf0c652a88) last updated 2016/10/12 21:23:05 (GMT +200)
  lib/ansible/modules/extras: (detached HEAD 6979330c1d) last updated 2016/10/12 21:23:05 (GMT +200)
  config file =
  configured module search path = Default w/o overrides
```

##### CONFIGURATION

No config file found; using defaults
##### OS / ENVIRONMENT

N/A

##### SUMMARY

When using custom `vars_plugins` each plugin is run for every host defined
in the inventory file. It should be run on the host subset only.

##### STEPS TO REPRODUCE

No playbook is needed

the `inventory` file is:

```
[g1]
host1
host2
host3

[g2]
host4
host5
host6
```

A simple supid `vars_plugins/test` (which can simulate one that consume
time) is:

```python
#!/usr/bin/env python

try:
    from __main__ import display
except ImportError:
    from ansible.utils.display import Display
    display = Display()

import time
    
class VarsModule(object):

    def __init__(self, inventory):
        self.inventory = inventory
                
    def get_host_vars(self, host, vault_password=None):
        time.sleep(2)
        display.display(host, color='cyan')

```

You can run:

```
time ansible all -i inventory -c local -m ping -v
```

and

```
time ansible g1 -i inventory -c local -m ping -v
```

and

```
time ansible host1 -i inventory -c local -m ping -v
```

##### EXPECTED RESULTS

- 1st command should be run in 12s
- 2nd command should be run in 6s
- 3rd command should be run in 2s

##### ACTUAL RESULTS

Each run takes about 13 or 13s to run.

```
No config file found; using defaults
host4
host5
host6
host1
host2
host3
host1 | SUCCESS => {
    "changed": false,
    "ping": "pong"
}

real	0m12.862s
user	0m0.631s
sys	0m0.210s
```

##### INVESTIGATIONS

In `lib/ansible/inventory/__init__.py` the `self.parse_inventory(host_list)`
(line 97) has no clue about any subset. Thus both `get_group_variables` and
`get_host_variables` are run over every single host defined in the inventory.

Is there a specific reason for that?

Wouldn't it be better to pass the subset to the `Inventory.__init__`
function in order to limit the `get_(host|group)_variables` to only the
required hosts?

**Comment by bcoca** [comment]
Then hostvars from hosts not targettted in play would not be available, as in this simple example:

```
- hosts: webservers
  tasks:
    - template: src=db_config.j2 dest=/etc/db_config
```

# db_config.j2
```
{% for db_server in groups['db_servers'] %}
db_server  username={{hostvars[db_server]['db_user']}} password={{hostvars[db_server]['db_password']}}
{% endfor %}
```

**Comment by renard** [comment]
Is it a real problem?

In my case (maybe I am misunderstanding something) I have a custom  var plugin that parses many yaml files (about 500) in several folders. The parsing time is quite long (+5s) for all files since the plugin is called for all hosts. 

I would prefer to have it run only on matched hosts. How can I do such a thing?

Thanks

**Comment by bcoca** [comment]
yes, its common usage, there are many other ways of making the hosts used dynamic so inventory must be loaded before the play runs. There is no way to do what you want.

**Comment by renard** [comment]
@bcoca Ok I see your point. I do not use the `hostvars[db_server]['db_user']` notation.

Let me know if something like a vars_subset_plugins is a good idea.

Thanks.

**Comment by robinro** [comment]
@renard Can your vars_plugin maybe cache the results? This is what we do to get the inventory/vars step quicker.

**Comment by bcoca** [comment]
closing as this is solved in 2.4, the vars plugins now run 'on demand', it now runs at 'task time' vs inventory compilation, file reads are cached as long as you use the loader object to read them in. It deals with hostvars by lazyloading them and triggering vars plugins if needed.
